File: miniapp/src/api/ability-points.ts
/**
 * 能力值积分系统 - Sprint 4.6
 * 与 backend /api/v1/ability-points 一致
 */
import request from '@/utils/request'

const PREFIX = '/api/v1/ability-points'

// ==================== 积分账户 ====================

/** 用户积分信息 */
export interface AbilityPointResponse {
  id: string
  userId: string
  totalPoints: number
  availablePoints: number
  level: number
  totalEarned: number
  totalSpent: number
  createdAt: string
  updatedAt: string
}

/** 获取我的积分信息 */
export function getMyPoints() {
  return request<AbilityPointResponse>({
    url: `${PREFIX}/my`,
    method: 'GET',
  }).then(data => {
    // 后端返回 snake_case 字段，映射到 camelCase
    if (data) {
      return {
        ...data,
        availablePoints: data.available_points ?? data.availablePoints,
        totalPoints: data.total_points ?? data.totalPoints,
        totalEarned: data.total_earned ?? data.totalEarned,
        totalSpent: data.total_spent ?? data.totalSpent,
        userId: data.user_id ?? data.userId,
        createdAt: data.created_at ?? data.createdAt,
        updatedAt: data.updated_at ?? data.updatedAt,
      } as AbilityPointResponse
    }
    return data
  })
}

// ==================== 积分流水 ====================

/** 积分流水记录 */
export interface PointTransactionResponse {
  id: string
  userId: string
  amount: number
  balanceAfter: number
  transactionType: string
  eventType?: string
  description?: string
  createdAt: string
}

/** 积分流水利表响应 */
export interface TransactionsListResponse {
  transactions: PointTransactionResponse[]
  total: number
}

/** 获取我的积分流水 */
export function getMyTransactions(limit = 50, offset = 0) {
  return request<TransactionsListResponse>({
    url: `${PREFIX}/my/transactions?limit=${limit}&offset=${offset}`,
    method: 'GET',
  }).then(res => {
    // 后端返回 snake_case 字段，映射到 camelCase
    if (res && res.transactions) {
      return {
        ...res,
        transactions: res.transactions.map((item: any) => ({
          ...item,
          transactionType: item.transaction_type ?? item.transactionType,
          eventType: item.event_type ?? item.eventType,
          createdAt: item.created_at ?? item.createdAt,
          balanceAfter: item.balance_after ?? item.balanceAfter,
          userId: item.user_id ?? item.userId,
        }))
      }
    }
    return res
  })
}

// ==================== 积分兑换 ====================

/** 兑换创建请求 */
export interface PointRedemptionCreate {
  rewardName: string
  rewardType: string
  pointsCost: number
  deliveryInfo?: string
  notes?: string
}

/** 兑换记录响应 */
export interface PointRedemptionResponse {
  id: string
  userId: string
  rewardName: string
  rewardType: string
  pointsCost: number
  status: 'pending' | 'approved' | 'completed' | 'rejected'
  deliveryInfo?: string
  notes?: string
  adminId?: string
  processedAt?: string
  rejectionReason?: string
  createdAt: string
  updatedAt: string
}

/** 兑换列表响应 */
export interface RedemptionsListResponse {
  redemptions: PointRedemptionResponse[]
  total: number
}

/** 创建积分兑换 */
export function createRedemption(data: PointRedemptionCreate) {
  return request<PointRedemptionResponse>({
    url: `${PREFIX}/redemptions`,
    method: 'POST',
    data,
  })
}

/** 获取我的兑换记录 */
export function getMyRedemptions(status?: string) {
  const url = status ? `${PREFIX}/redemptions?status=${status}` : `${PREFIX}/redemptions`
  return request<RedemptionsListResponse>({
    url,
    method: 'GET',
  })
}

// ==================== 积分事件 ====================

/** 积分事件项 */
export interface PointEvent {
  eventType: string
  points: number
}

/** 积分事件列表响应 */
export interface EventsListResponse {
  events: PointEvent[]
  total: number
}

/** 获取积分事件列表 */
export function getPointEvents() {
  return request<EventsListResponse>({
    url: `${PREFIX}/events`,
    method: 'GET',
  })
}

File: backend/app/schemas/ability_points.py
"""Ability Points Schemas - 能力值系统数据验证（Sprint 4.6）"""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


# ============== Ability Point Response ==============
class AbilityPointResponse(BaseModel):
    """用户积分信息响应"""
    id: str
    user_id: str
    total_points: int
    available_points: int
    level: int
    total_earned: int
    total_spent: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ============== Transaction Schemas ==============
class AbilityPointTransactionResponse(BaseModel):
    """积分流水响应"""
    id: str
    user_id: str
    amount: int
    balance_after: int
    transaction_type: str
    event_type: Optional[str] = None
    description: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ============== Redemption Schemas ==============
class PointRedemptionCreate(BaseModel):
    """创建积分兑换请求"""
    reward_name: str = Field(..., min_length=1, max_length=100, description="奖励名称")
    reward_type: str = Field(..., description="奖励类型: coupon/gift/vip/etc.")
    points_cost: int = Field(..., gt=0, description="消耗积分")
    delivery_info: Optional[str] = Field(None, description="配送信息 JSON")
    notes: Optional[str] = Field(None, description="备注")


class PointRedemptionUpdate(BaseModel):
    """更新兑换状态（管理员）"""
    status: str = Field(..., pattern="^(pending|approved|completed|rejected)$")
    rejection_reason: Optional[str] = Field(None, description="拒绝原因")


class PointRedemptionResponse(BaseModel):
    """兑换记录响应"""
    id: str
    user_id: str
    reward_name: str
    reward_type: str
    points_cost: int
    status: str
    delivery_info: Optional[str] = None
    notes: Optional[str] = None
    admin_id: Optional[str] = None
    processed_at: Optional[datetime] = None
    rejection_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ============== Admin Schemas ==============
class AdminRedemptionListResponse(BaseModel):
    """管理员兑换列表响应"""
    redemptions: List[PointRedemptionResponse]
    total: int
    pending_count: int


# ============== Event Hook Schema ==============
class PointEvent(BaseModel):
    """积分事件（用于hooks系统）"""
    user_id: str
    event_type: str = Field(..., description="事件类型")
    points: int = Field(..., description="积分变动（正为获得，负为消费）")
    reference_id: Optional[str] = Field(None, description="关联记录ID")
    reference_type: Optional[str] = Field(None, description="关联记录类型")
    description: Optional[str] = Field(None, description="描述")
